data/model/storage.py

```diff
@@ -69,43 +69,25 @@ def add_storage_placement(storage, location_name):
         pass
 
 
-def _orphaned_storage_query(candidate_ids):
+def _is_storage_orphaned(candidate_id):
     """
-    Returns the subset of the candidate ImageStorage IDs representing storages that are no longer
-    referenced by images.
+    Returns the whether the given candidate storage ID is orphaned. Must be executed
+    under a transaction.
     """
-    # Issue a union query to find all storages that are still referenced by a candidate storage. This
-    # is much faster than the group_by and having call we used to use here.
-    nonorphaned_queries = []
-    for counter, candidate_id in enumerate(candidate_ids):
-        query_alias = "q{0}".format(counter)
-
-        # TODO: remove the join with Image once fully on the OCI data model.
-        storage_subq = (
-            ImageStorage.select(ImageStorage.id)
-            .join(Image)
-            .where(ImageStorage.id == candidate_id)
-            .limit(1)
-            .alias(query_alias)
-        )
-
-        nonorphaned_queries.append(ImageStorage.select(SQL("*")).from_(storage_subq))
-
-        manifest_storage_subq = (
-            ImageStorage.select(ImageStorage.id)
-            .join(ManifestBlob)
-            .where(ImageStorage.id == candidate_id)
-            .limit(1)
-            .alias(query_alias)
-        )
+    with ensure_under_transaction():
+        try:
+            ManifestBlob.get(blob=candidate_id)
+            return False
+        except ManifestBlob.DoesNotExist:
+            pass
 
-        nonorphaned_queries.append(ImageStorage.select(SQL("*")).from_(manifest_storage_subq))
+        try:
+            Image.get(storage=candidate_id)
+            return False
+        except Image.DoesNotExist:
+            pass
```

Comment:
I wonder if unioning these two queries and using `COUNT()` to see if the number of results are greater than zero is less overhead for the database (e.g. is it cheaper than opening another connection).

Reply:
Well, it doesn't open another connection, but it does run another query.

We could LEFT OUTER JOIN and then just retrieve the first result (if any), I suppose, but then we're holding two tables at once instead of a simple index lookup.

Reply:
In the grand scheme of things, doing so would be premature optimization. If you were running many, many instances of Quay, you'd benefit from having GC reduce the number of queries it makes, but that's always going to be a drop in the bucket compared to the queries made by the API.